### ISSUE TITLE
Hard/Soft Reset missing from Libretro

### ISSUE THREAD

**Issue by Sonia-7** [TASEmulators/BizHawk#3482]
~~I decided to make this a 2-in-1 issue instead of making them separated, since I believe they're closely related.~~ CNR bug report, removed to leave feature request
Anyway, I think the title already says all. It'd be nice to fix this before 2.9 gets released.

**Comment by YoshiRulz** [comment]
Did this core work in 2.8? The Libretro subsystem has been rewritten in 2.9.

**Comment by CasualPokePlayer** [comment]
The libretro API doesn't define hard vs soft reset (`retro_reset` is parameterless)

Also I recall this core working fine here before, the crash might just be the fault of the native side / side we don't control / maybe works on retroarch because they don't follow their own specs half the time.

**Comment by Sonia-7** [comment]
@YoshiRulz: I don''t think Hard/Soft reset was ever a thing. But reboots definitely didn't crash the emu in 2.8.

**Comment by CasualPokePlayer** [comment]
Which one? Upstream's or libretro's fork?

**Comment by Sonia-7** [comment]
Casual: Upstream.
Yoshi: That laugh looks meaningless to me. What was the point of that?

**Comment by CasualPokePlayer** [comment]
Compiled myself with upstream's Makefile.libretro, no crash on reboot core.

Note: I'm using `Gargoyles (USA).md` here for testing

**Comment by YoshiRulz** [comment]
Their build scripts mustn't be programmed very well if they only work sometimes. (I'm being a bit hypocritical though, Lua not working in yesterday's RC1 was because of my mistake.)

**Comment by Sonia-7** [comment]
Ok, so I figured out what the problem was. In the "libretro_core_options.h" file, there's a "genesis_plus_gx_region_detect" section. If you edit that into "ntsc-u" or "pal" or "ntsc-j" and then compile, it will start crashing. Only "auto" (the default) works fine. This only became an issue after the libretro rewrite.

**Comment by CasualPokePlayer** [comment]
The crash occurs here: https://github.com/ekeeke/Genesis-Plus-GX/blob/cea418ece8152520faf1a9aea2d17a89906a6dc7/core/sound/blip_buf.c#L182

`blip_set_rates` called here is from https://github.com/ekeeke/Genesis-Plus-GX/blob/3cf83e7557a7869fbe312ca059355d2cc76058e6/core/system.c#L131 https://github.com/ekeeke/Genesis-Plus-GX/blob/6ec9ac2729b6bd4b67b61e3c5ccd4683ef6d995f/libretro/libretro.c#L1392

snd.blips[0] is NULL at this point.

snd.blips would have its blips freed and cleared in `audio_shutdown` in `retro_unload_game` here: https://github.com/ekeeke/Genesis-Plus-GX/blob/6ec9ac2729b6bd4b67b61e3c5ccd4683ef6d995f/libretro/libretro.c#L3375 (which is likely where the difference came from as the old libretro system didn't call retro_unload_game at all, while the new one does).

snd.blips normally has its blips set in `audio_init`, which could be called here https://github.com/ekeeke/Genesis-Plus-GX/blob/6ec9ac2729b6bd4b67b61e3c5ccd4683ef6d995f/libretro/libretro.c#L3327 or here https://github.com/ekeeke/Genesis-Plus-GX/blob/6ec9ac2729b6bd4b67b61e3c5ccd4683ef6d995f/libretro/libretro.c#L1973

Notice how the latter is in the same function as the `audio_set_rate` call crashing, but it's *after* the `audio_set_rate` call, so too late.

It doesn't crash on the very first run however since `system_hw` is 0 at that point, so it skips the `audio_set_rate` call. `retro_unload_game` does not clear `system_hw` so it calls it on the reboot core.

tl;dr this is a bug on the core's side, not our side. Go report this crash there.
